### DISCUSSION TITLE
dwl wayland alacritty 里突然中文输入法不工作了

### DISCUSSION THREAD

**Discussion by guyuming76** [fcitx/fcitx5#512] [Q&A] [answered]
我前段时间在 dwl alacritty 里通过 wayland 用fcitx5 拼音输入法一直是好的：  https://gitee.com/guyuming76/dwl/tree/PR235_8/ 

但突然这两天 ctrl+space 就无法切换到中文了，我在 dwl 这边调试代码发现 input_method->keyboard_grab 返回空，我不知道是dwl这边的问题，还是 wlroots 的问题，还是fcitx5. 

https://pastebin.com/FM9a2tty  是 fcitx5 --verbose "*=5" 的输出，期间我用 ctrl+space 在 alacritty 里切换输入法，但没反应。 这个问题我改从日志那里找线索呢，或者用 gdb 调哪些断点来判断？

https://pastebin.com/Eks8JUYq 是fcitx5-diagnose 结果

**Comment by wengxt** [comment, marked as answer]
建议用 WAYLAND_DEBUG=1 来看两者之间的交互

这里看到有提示，至少已经创建了对应的连接对象
  IC [af39a8c989ed42d4af4cdc0904e85c99] program: frontend:wayland_v2 cap:52 focus:0

另外 keyboard_grab  只有在有 text-input 窗口激活的时候才有，Debuginfo 提示的 focus 应当为 1

**Reply by guyuming76** [reply]
多谢，我拿之前能工作的commit 一个一个比较，原来自己代码里犯了个低级错误，#ifdef 后面的大小写搞错了：https://gitee.com/guyuming76/dwl/commit/7ad12a9aaa8aa561a5dd6e469917f0415dd9739b 

另外这里的代码为了考虑 XWAYLAND, 每次enable, disable text_input 的时候，都把 Input_Method 也给 activate 和 deactivate 掉了。 必须这样吗？有没有别的办法？ 

如果不考虑 XWAYLAND, 我是不是可以只在 Input_Method 创建之后 activate 一次，以后 text_input enable 的时候就不用 activate input_method 了？

**Reply by wengxt** [reply]
不行，activate 和 deactivate 表示了客户端程序获得和失去焦点。在这些事件发生的时候输入法一般来说还是要进行一些操作的（一般是请看状态），还有 fcitx 的获得焦点时显示当前输入法也是需要这些事件正确处理。

另外，正因为他们是获得焦点相关的，一旦获得焦点输入法就会进行 key grab，意味着按键都要经过输入法，而原本是不需要的。除了性能会受到影响（按键经过输入法，又由输入法再注入），还会影响某些按键的处理，本来不应当被输入法使用的按键可能也会被处理。
